neps/nep-0509.md

```diff
@@ -0,0 +1,592 @@
+  * Previously there was no limit on receipt size. Set to 4 MiB, might be reduced to 1.5 MiB in the future to match the transaction limit.
+* `combined_transactions_size_limit - 4 MiB`
+  * Hard limit on total size of transactions from this and previous chunk. `ChunkStateWitness` contains transactions from two chunks, this limit applies to the sum of their sizes.
+* `new_transactions_validation_state_size_soft_limit - 500 KiB`
+  * Validating new transactions generates storage proof (recorded trie nodes), which has to be limited. Once transaction validation generates more storage proof than this limit, the chunk producer stops adding new transactions to the chunk.
+* `per_receipt_storage_proof_size_limit - 4 MB`
+  * Executing a receipt generates storage proof. A single receipt is allowed to generate at most 4 MB of storage proof. This is a hard limit, receipts which generate more than that will fail.
+* `main_storage_proof_size_soft_limit - 3 MB`
+  * This is a limit on the total size of storage proof generated by receipts in one chunk. Once receipts generate more storage proof than this limit, the chunk producer stops processing receipts and moves the rest to the delayed queue.
+  * It's a soft limit, which means that the total size of storage proof could reach 7 MB (2.99MB + one receipt which generates 4MB of storage proof)
+  * Due to implementation details it's hard to find the exact amount of storage proof generated by a receipt, so an upper bound estimation is used instead. This upper bound assumes that every removal generates additional 2000 bytes of storage proof, so receipts which perform a lot of trie removals might be limited more than theoretically applicable.
+* `outgoing_receipts_usual_size_limit - 100 KiB`
+  * Limit on the size of outgoing receipts to another shard. Needed to keep the size of `source_receipt_proofs` small.
+  * On most block heights a shard isn't allowed to send receipts larger than 100 KiB to another shard.
+* `outgoing_receipts_big_size_limit - 4.5 MiB`
+  * On every block height there's one special "allowed shard" which is allowed to send larger receipts, up to 4.5 MiB in total.
+  * A receiving shard will receive receipts from `num_shards - 1` shards using the usual limit and one shard using the big limit.
+  * The "allowed shard" is the same shard as in cross-shard congestion control. It's chosen in a round-robin fashion, at height 1 the special shard is 0, at height 2 it's 1 and so on.
+
+In total that gives 4 MiB + 500 KiB + 7MB + 5*100 KiB + 4.5 MiB ~= 16 MiB of maximum witness size. Possibly a little more on missing chunks.
+
+### New limits breaking contracts
+
+The new limits will break some existing contracts (for example all transctions larger than 1.5 MiB). This is sad, but it's necessary. Stateless validation uses much more network bandwitch than the previous approach, as it has to send over all state on each chunk application. Because network bandwitch is limited, stateless validation
```

Comment:
```suggestion
The new limits will break some existing contracts (for example, all transactions larger than 1.5 MiB). This is sad, but it's necessary. Stateless validation uses much more network bandwidth than the previous approach, as it has to send over all states on each chunk application. Because network bandwidth is limited, stateless validation
```